core/src/main/java/hudson/model/Actionable.java

```diff
@@ -120,31 +133,124 @@
+        List<Action> current = getOrCreateActions();
+        boolean found = false;
         for (Action a2 : current) {
-            if (a2.getClass() == a.getClass()) {
+            if (!found && a.equals(a2)) {
+                found = true;
+            } else  if (a2.getClass() == a.getClass()) {
                 old.add(a2);
             }
         }
         current.removeAll(old);
-        addAction(a);
+        if (!found) {
+            addAction(a);
+        }
+    }
+
+    /**
+     * Remove an action. Note calls to {@link #getAllActions()} that happen before calls to this method may not see the
+     * update.
+     *
+     * @param a an action to remove (if {@code null} then this will be a no-op)
+     * @return {@code true} if this actions changed as a result of the call
+     * @since FIXME
+     */
+    @SuppressWarnings("deprecation")
+    public boolean removeAction(@Nullable Action a) {
+        if (a == null) {
+            return false;
+        }
+        // CopyOnWriteArrayList does not support Iterator.remove, so need to do it this way:
+        return getOrCreateActions().removeAll(Collections.singleton(a));
+    }
+
+    /**
+     * Removes any actions of the specified type. Note calls to {@link #getAllActions()} that happen before calls to
+     * this method may not see the update.
+     *
+     * @param clazz the type of actions to remove
+     * @return {@code true} if this actions changed as a result of the call
+     * @since FIXME
+     */
+    @SuppressWarnings({"ConstantConditions","deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public boolean removeActions(@Nonnull Class<? extends Action> clazz) {
+        if (clazz == null) {
+            throw new IllegalArgumentException("Action type must be non-null");
+        }
+        // CopyOnWriteArrayList does not support Iterator.remove, so need to do it this way:
+        List<Action> old = new ArrayList<Action>();
+        List<Action> current = getOrCreateActions();
+        for (Action a : current) {
+            if (clazz.isInstance(a)) {
+                old.add(a);
+            }
+        }
+        return current.removeAll(old);
+    }
+
+    /**
+     * Replaces any actions of the specified type by the supplied action. Note calls to {@link #getAllActions()} that
+     * happen before calls to this method may not see the update.
+     *
+     * @param clazz the type of actions to replace (note that the action you are replacing this with need not extend
+     *              this class)
+     * @param a     the action to replace with
+     * @return {@code true} if this actions changed as a result of the call
+     * @since FIXME
+     */
+    @SuppressWarnings({"ConstantConditions", "deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public boolean replaceActions(@Nonnull Class<? extends Action> clazz, Action a) {
+        if (clazz == null) {
```

Comment:
🐛 missing check/annotation for `a`, which is actually Nonnull in this code. Bug only since other methods check it